File: test/test_hide.py
import io

from branchless.hide import hide, unhide
from branchless.smartlog import smartlog
from helpers import Git, compare


def test_hide_commit(git: Git) -> None:
    git.init_repo()
    git.detach_head()
    git.commit_file(name="test1", time=1)
    git.run("checkout", ["master"])
    git.detach_head()
    git.commit_file(name="test2", time=2)

    with io.StringIO() as out:
        assert smartlog(out=out) == 0
        compare(
            actual=out.getvalue(),
            expected="""\
O f777ecc9 (master) create initial.txt
|\\
| o 62fc20d2 create test1.txt
|
@ fe65c1fe create test2.txt
""",
        )

    with io.StringIO() as out:
        assert hide(out=out, hashes=["62fc20d2"]) == 0
        compare(
            actual=out.getvalue(),
            expected="""\
Hid commit: 62fc20d2 create test1.txt
To unhide this commit, run: git unhide 62fc20d2
""",
        )

    with io.StringIO() as out:
        assert smartlog(out=out) == 0
        compare(
            actual=out.getvalue(),
            expected="""\
O f777ecc9 (master) create initial.txt
|
@ fe65c1fe create test2.txt
""",
        )


def test_hide_bad_commit(git: Git) -> None:
    git.init_repo()
    with io.StringIO() as out:
        assert hide(out=out, hashes=["abc123"]) == 1
        compare(
            actual=out.getvalue(),
            expected="""\
Commit not found: abc123
""",
        )


def test_hide_transitive(git: Git) -> None:
    git.init_repo()
    git.detach_head()
    git.commit_file(name="test1", time=1)
    git.commit_file(name="test2", time=2)
    git.commit_file(name="test3", time=3)
    git.run("checkout", ["master"])

    with io.StringIO() as out:
        smartlog(out=out)
        compare(
            actual=out.getvalue(),
            expected="""\
@ f777ecc9 (master) create initial.txt
|
o 62fc20d2 create test1.txt
|
o 96d1c37a create test2.txt
|
o 70deb1e2 create test3.txt
""",
        )

    with io.StringIO() as out:
        assert hide(out=out, hashes=["96d1c37a"]) == 0

    with io.StringIO() as out:
        smartlog(out=out)
        compare(
            actual=out.getvalue(),
            expected="""\
@ f777ecc9 (master) create initial.txt
|
o 62fc20d2 create test1.txt
""",
        )


def test_hide_already_hidden_commit(git: Git) -> None:
    git.init_repo()
    git.detach_head()
    git.commit_file(name="test1", time=1)

    with io.StringIO() as out:
        assert hide(out=out, hashes=["62fc20d2"]) == 0

    with io.StringIO() as out:
        assert hide(out=out, hashes=["62fc20d2"]) == 0
        compare(
            actual=out.getvalue(),
            expected="""\
Hid commit: 62fc20d2 create test1.txt
(It was already hidden, so this operation had no effect.)
To unhide this commit, run: git unhide 62fc20d2
""",
        )


def test_hide_current_commit(git: Git) -> None:
    git.init_repo()
    git.detach_head()
    git.commit_file(name="test", time=1)

    with io.StringIO() as out:
        assert hide(out=out, hashes=["HEAD"]) == 0

    with io.StringIO() as out:
        assert smartlog(out=out) == 0
        compare(
            expected=out.getvalue(),
            actual="""\
O f777ecc9 (master) create initial.txt
|
% 3df4b935 create test.txt
""",
        )


def test_hidden_commit_with_head_as_child(git: Git) -> None:
    git.init_repo()
    git.detach_head()
    git.commit_file(name="test1", time=1)
    git.commit_file(name="test2", time=2)
    git.commit_file(name="test3", time=3)
    git.run("checkout", ["HEAD^"])

    with io.StringIO() as out:
        assert hide(out=out, hashes=["HEAD^"]) == 0
        assert hide(out=out, hashes=["70deb1e2"]) == 0

    with io.StringIO() as out:
        assert smartlog(out=out) == 0
        compare(
            actual=out.getvalue(),
            expected="""\
O f777ecc9 (master) create initial.txt
|
x 62fc20d2 create test1.txt
|
@ 96d1c37a create test2.txt
""",
        )


def test_hide_master_commit_with_hidden_children(git: Git) -> None:
    git.init_repo()
    git.commit_file(name="test1", time=1)
    git.commit_file(name="test2", time=2)
    git.detach_head()
    git.commit_file(name="test3", time=3)
    git.run("checkout", ["master"])
    git.commit_file(name="test4", time=4)
    git.commit_file(name="test5", time=5)
    git.run("reflog", ["delete", "HEAD@{1}"])

    with io.StringIO() as out:
        assert hide(out=out, hashes=["70deb1e2"]) == 0

    with io.StringIO() as out:
        assert smartlog(out=out) == 0
        compare(
            actual=out.getvalue(),
            expected="""\
:
@ 20230db7 (master) create test5.txt
""",
        )


def test_branches_always_visible(git: Git) -> None:
    git.init_repo()
    git.detach_head()
    git.commit_file(name="test1", time=1)
    git.commit_file(name="test2", time=2)
    git.run("branch", ["test"])
    git.run("checkout", ["master"])

    with io.StringIO() as out:
        assert hide(out=out, hashes=["test", "test^"]) == 0

    with io.StringIO() as out:
        assert smartlog(out=out) == 0
        compare(
            actual=out.getvalue(),
            expected="""\
@ f777ecc9 (master) create initial.txt
|
x 62fc20d2 create test1.txt
|
x 96d1c37a (test) create test2.txt
""",
        )

    git.run("branch", ["-D", "test"])

    with io.StringIO() as out:
        assert smartlog(out=out) == 0
        compare(
            actual=out.getvalue(),
            expected="""\
@ f777ecc9 (master) create initial.txt
""",
        )


def test_unhide(git: Git) -> None:
    git.init_repo()
    git.detach_head()
    git.commit_file(name="test1", time=1)
    git.commit_file(name="test2", time=2)
    git.run("checkout", ["master"])

    with io.StringIO() as out:
        assert unhide(out=out, hashes=["96d1c37"]) == 0
        compare(
            actual=out.getvalue(),
            expected="""\
Unhid commit: 96d1c37a create test2.txt
(It was not hidden, so this operation had no effect.)
To hide this commit, run: git hide 96d1c37a
""",
        )

    with io.StringIO() as out:
        assert hide(out=out, hashes=["96d1c37"]) == 0

    with io.StringIO() as out:
        assert smartlog(out=out) == 0
        compare(
            actual=out.getvalue(),
            expected="""\
@ f777ecc9 (master) create initial.txt
|
o 62fc20d2 create test1.txt
""",
        )

    with io.StringIO() as out:
        assert unhide(out=out, hashes=["96d1c37"]) == 0
        compare(
            actual=out.getvalue(),
            expected="""\
Unhid commit: 96d1c37a create test2.txt
To hide this commit, run: git hide 96d1c37a
""",
        )

    with io.StringIO() as out:
        assert smartlog(out=out) == 0
        compare(
            actual=out.getvalue(),
            expected="""\
@ f777ecc9 (master) create initial.txt
|
o 62fc20d2 create test1.txt
|
o 96d1c37a create test2.txt
""",
        )

File: tests/test_hide.rs
use branchless::testing::{with_git, GitRunOptions};

#[test]
fn test_hide_commit() -> anyhow::Result<()> {
    with_git(|git| {
        git.init_repo()?;
        git.detach_head()?;
        let test1_oid = git.commit_file("test1", 1)?;
        git.run(&["checkout", "master"])?;
        git.detach_head()?;
        git.commit_file("test2", 2)?;

        {
            let (stdout, _stderr) = git.run(&["smartlog"])?;
            insta::assert_snapshot!(stdout, @r###"
            O f777ecc9 (master) create initial.txt
            |\
            | o 62fc20d2 create test1.txt
            |
            @ fe65c1fe create test2.txt
            "###);
        }

        {
            let (stdout, _stderr) = git.run(&["hide", &test1_oid.to_string()])?;
            insta::assert_snapshot!(stdout, @r###"
            Hid commit: 62fc20d2 create test1.txt
            To unhide this commit, run: git unhide 62fc20d2
            "###);
        }

        {
            let (stdout, _stderr) = git.run(&["smartlog"])?;
            insta::assert_snapshot!(stdout, @r###"
            O f777ecc9 (master) create initial.txt
            |
            @ fe65c1fe create test2.txt
            "###);
        }

        Ok(())
    })
}

#[test]
fn test_hide_bad_commit() -> anyhow::Result<()> {
    with_git(|git| {
        git.init_repo()?;

        {
            let (stdout, _stderr) = git.run_with_options(
                &["hide", "abc123"],
                &GitRunOptions {
                    expected_exit_code: 1,
                    ..Default::default()
                },
            )?;
            insta::assert_snapshot!(stdout, @"Commit not found: abc123");
        }

        Ok(())
    })
}

#[test]
fn test_hide_already_hidden_commit() -> anyhow::Result<()> {
    with_git(|git| {
        git.init_repo()?;
        git.detach_head()?;
        let test1_oid = git.commit_file("test1", 1)?;

        git.run(&["hide", &test1_oid.to_string()])?;
        {
            let (stdout, _stderr) = git.run(&["hide", &test1_oid.to_string()])?;
            insta::assert_snapshot!(stdout, @r###"
            Hid commit: 62fc20d2 create test1.txt
            (It was already hidden, so this operation had no effect.)
            To unhide this commit, run: git unhide 62fc20d2
            "###);
        }

        Ok(())
    })
}

#[test]
fn test_hide_current_commit() -> anyhow::Result<()> {
    with_git(|git| {
        git.init_repo()?;
        git.detach_head()?;
        git.commit_file("test", 1)?;
        git.run(&["hide", "HEAD"])?;

        {
            let (stdout, _stderr) = git.run(&["smartlog"])?;
            insta::assert_snapshot!(stdout, @r###"
            O f777ecc9 (master) create initial.txt
            |
            % 3df4b935 create test.txt
            "###);
        }

        Ok(())
    })
}

#[test]
fn test_hidden_commit_with_head_as_child() -> anyhow::Result<()> {
    with_git(|git| {
        git.init_repo()?;
        git.detach_head()?;
        let test1_oid = git.commit_file("test1", 1)?;
        let test2_oid = git.commit_file("test2", 2)?;
        let test3_oid = git.commit_file("test3", 3)?;
        git.run(&["checkout", &test2_oid.to_string()])?;

        git.run(&["hide", &test1_oid.to_string()])?;
        git.run(&["hide", &test3_oid.to_string()])?;

        {
            let (stdout, _stderr) = git.run(&["smartlog"])?;
            insta::assert_snapshot!(stdout, @r###"
            O f777ecc9 (master) create initial.txt
            |
            x 62fc20d2 create test1.txt
            |
            @ 96d1c37a create test2.txt
            "###);
        }

        Ok(())
    })
}

#[test]
fn test_hide_master_commit_with_hidden_children() -> anyhow::Result<()> {
    with_git(|git| {
        git.init_repo()?;
        git.commit_file("test1", 1)?;
        git.commit_file("test2", 2)?;
        git.detach_head()?;
        let test3_oid = git.commit_file("test3", 3)?;
        git.run(&["checkout", "master"])?;
        git.commit_file("test4", 4)?;
        git.commit_file("test5", 5)?;

        git.run(&["hide", &test3_oid.to_string()])?;
        {
            let (stdout, _stderr) = git.run(&["smartlog"])?;
            insta::assert_snapshot!(stdout, @r###"
            :
            @ 20230db7 (master) create test5.txt
            "###);
        }

        Ok(())
    })
}

#[test]
fn test_branches_always_visible() -> anyhow::Result<()> {
    with_git(|git| {
        git.init_repo()?;
        git.detach_head()?;
        git.commit_file("test1", 1)?;
        git.commit_file("test2", 2)?;
        git.run(&["branch", "test"])?;
        git.run(&["checkout", "master"])?;

        git.run(&["hide", "test", "test^"])?;
        {
            let (stdout, _stderr) = git.run(&["smartlog"])?;
            insta::assert_snapshot!(stdout, @r###"
            @ f777ecc9 (master) create initial.txt
            |
            x 62fc20d2 create test1.txt
            |
            x 96d1c37a (test) create test2.txt
            "###);
        }

        git.run(&["branch", "-D", "test"])?;
        {
            let (stdout, _stderr) = git.run(&["smartlog"])?;
            insta::assert_snapshot!(stdout, @"@ f777ecc9 (master) create initial.txt
");
        }

        Ok(())
    })
}

#[test]
fn test_unhide() -> anyhow::Result<()> {
    with_git(|git| {
        git.init_repo()?;
        git.detach_head()?;
        git.commit_file("test1", 1)?;
        let test2_oid = git.commit_file("test2", 2)?;
        git.run(&["checkout", "master"])?;

        {
            let (stdout, _stderr) = git.run(&["unhide", &test2_oid.to_string()])?;
            insta::assert_snapshot!(stdout, @r###"
            Unhid commit: 96d1c37a create test2.txt
            (It was not hidden, so this operation had no effect.)
            To hide this commit, run: git hide 96d1c37a
            "###);
        }

        git.run(&["hide", &test2_oid.to_string()])?;
        {
            let (stdout, _stderr) = git.run(&["smartlog"])?;
            insta::assert_snapshot!(stdout, @r###"
            @ f777ecc9 (master) create initial.txt
            |
            o 62fc20d2 create test1.txt
            "###);
        }

        {
            let (stdout, _stderr) = git.run(&["unhide", &test2_oid.to_string()])?;
            insta::assert_snapshot!(stdout, @r###"
            Unhid commit: 96d1c37a create test2.txt
            To hide this commit, run: git hide 96d1c37a
            "###);
        }

        {
            let (stdout, _stderr) = git.run(&["smartlog"])?;
            insta::assert_snapshot!(stdout, @r###"
            @ f777ecc9 (master) create initial.txt
            |
            o 62fc20d2 create test1.txt
            |
            o 96d1c37a create test2.txt
            "###);
        }

        Ok(())
    })
}

#[test]
fn test_hide_recursive() -> anyhow::Result<()> {
    with_git(|git| {
        git.init_repo()?;
        git.detach_head()?;
        git.commit_file("test1", 1)?;
        let test2_oid = git.commit_file("test2", 2)?;
        git.commit_file("test3", 3)?;
        git.run(&["checkout", "master"])?;

        {
            let (stdout, _stderr) = git.run(&["smartlog"])?;
            insta::assert_snapshot!(stdout, @r###"
            @ f777ecc9 (master) create initial.txt
            |
            o 62fc20d2 create test1.txt
            |
            o 96d1c37a create test2.txt
            |
            o 70deb1e2 create test3.txt
            "###);
        }

        {
            let (stdout, _stderr) = git.run(&["hide", "-r", &test2_oid.to_string()])?;
            insta::assert_snapshot!(stdout, @r###"
            Hid commit: 96d1c37a create test2.txt
            To unhide this commit, run: git unhide 96d1c37a
            Hid commit: 70deb1e2 create test3.txt
            To unhide this commit, run: git unhide 70deb1e2
            "###);
        }

        {
            let (stdout, _stderr) = git.run(&["smartlog"])?;
            insta::assert_snapshot!(stdout, @r###"
            @ f777ecc9 (master) create initial.txt
            |
            o 62fc20d2 create test1.txt
            "###);
        }

        {
            let (stdout, _stderr) = git.run(&["unhide", "-r", &test2_oid.to_string()])?;
            insta::assert_snapshot!(stdout, @r###"
            Unhid commit: 96d1c37a create test2.txt
            To hide this commit, run: git hide 96d1c37a
            Unhid commit: 70deb1e2 create test3.txt
            To hide this commit, run: git hide 70deb1e2
            "###);
        }

        {
            let (stdout, _stderr) = git.run(&["smartlog"])?;
            insta::assert_snapshot!(stdout, @r###"
            @ f777ecc9 (master) create initial.txt
            |
            o 62fc20d2 create test1.txt
            |
            o 96d1c37a create test2.txt
            |
            o 70deb1e2 create test3.txt
            "###);
        }

        Ok(())
    })
}
